### PR TITLE
fix: ClassFormatError when generating writers for array-typed fields (#4009)

### LINKED ISSUE #4009: [BUG] 升级到fastjson 2.0.57报错
### 问题描述
*简要描述您碰到的问题。*

### 环境信息
*请填写以下信息：*

 - OS信息：  win11
 - JDK信息： OpenJDK 17.0.2 
 - 版本信息：fastjson2.0.57
 

### 重现步骤
*如何操作可以重现该问题：*

1. 使用 JSON.toJSON方法
2. 输入 `new DefinDto()`  数据
3. 出现 `Exception in thread "main" java.lang.ClassFormatError: Illegal class name "com/alibaba/fastjson2/writer/OWG_5_0_Class[]" in class file com/alibaba/fastjson2/writer/OWG_5_0_Class[]
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
	at com.alibaba.fastjson2.util.DynamicClassLoader.defineClassPublic(DynamicClassLoader.java:221)
	at com.alibaba.fastjson2.writer.ObjectWriterCreatorASM.jitWriter(ObjectWriterCreatorASM.java:576)
	at com.alibaba.fastjson2.writer.ObjectWriterCreatorASM.createObjectWriter(ObjectWriterCreatorASM.java:447)
	at com.alibaba.fastjson2.writer.ObjectWriterProvider.getObjectWriterInternal(ObjectWriterProvider.java:464)
	at com.alibaba.fastjson2.writer.ObjectWriterProvider.getObjectWriter(ObjectWriterProvider.java:340)
	at com.alibaba.fastjson2.JSONFactory.getObjectWriter(JSONFactory.java:493)
	at com.alibaba.fastjson2.writer.ObjectWriterAdapter.toJSONObject(ObjectWriterAdapter.java:681)
	at com.alibaba.fastjson2.writer.ObjectWriterAdapter.toJSONObject(ObjectWriterAdapter.java:610)
	at com.alibaba.fastjson2.writer.ObjectWriterAdapter.toJSONObject(ObjectWriterAdapter.java:685)
	at com.alibaba.fastjson2.JSON.toJSON(JSON.java:3926)
	at com.alibaba.fastjson2.JSON.toJSON(JSON.java:3898)
	at com.alibaba.fastjson.JSON.toJSON(JSON.java:2309)` 错误
```java
    OapiRobotSendRequest request = new OapiRobotSendRequest();
    Object json = JSON.toJSON(request);
```

### 期待的正确结果
*对您期望发生的结果进行清晰简洁的描述。*

### 相关日志输出
*请复制并粘贴任何相关的日志输出。*

#### 附加信息

Image

### PR DESCRIPTION
**Pull request by wenshao**
## Problem

`JSON.toJSON(bean)` throws `ClassFormatError` when the bean has an array-typed field:

```java
public class BeanWithClassArray {
    private String name;
    private Class<?>[] classes;
    // getters / setters
}

JSON.toJSON(bean);
```

```
java.lang.ClassFormatError: Illegal class name "com/alibaba/fastjson2/writer/OWG_5_0_Class[]"
        in class file com/alibaba/fastjson2/writer/OWG_5_0_Class[]
```

## Root cause

`ObjectWriterCreatorASM` embeds `objectClass.getSimpleName()` in the name of the class it generates:

```java
String className = "OWG_" + seed.incrementAndGet() + "_" + fieldWriters.size()
        + (objectClass == null ? "" : ("_" + objectClass.getSimpleName()));
```

For an array type the simple name ends in `[]` — `Class[]` — and `[` is not legal in a class name, so `defineClass` rejects the generated bytes.

## Fix

Sanitize the simple name before embedding it: array notation becomes a suffix (`Class[]` → `ClassArray`), and any other character that is not valid in a class name becomes an underscore.

The helper lives in `com.alibaba.fastjson2.internal.asm.ASMUtils`, which both `ObjectWriterCreatorASM` and `ObjectReaderCreatorASM` already static-import, so the two cannot drift apart:

```java
public static String sanitizeClassName(String simpleName)
```

### Scope

The writer is the only path that could produce an illegal name. The reader call sites are defensive symmetry, verified rather than assumed:

```
reader(Class[]) -> com.alibaba.fastjson2.reader.ObjectReaderAdapter
writer(Class[]) -> com.alibaba.fastjson2.writer.OWG_1_0_ClassArray
```

`ObjectReaderCreatorASM` never generates a class for an array type — it returns `ObjectReaderAdapter`, so the `ORG_` path is not reached. Its other call site (`VBACG_`/`VCACG_` in `createValueConsumer0`) bails out earlier still, since it requires a default constructor and a public class, and an array type has neither.

## Tests

- `Issue4009Test` — serialization and round-trip for `Class[]` and `int[][]` fields.
- `ASMUtilsTest` — unit coverage for `sanitizeClassName`: null/empty, unchanged names, arrays including `int[][]`, special characters, and the combined `My$Class[]`.

## Verification

Revert-probe — removing `sanitizeClassName` from `ObjectWriterCreatorASM` makes the integration test fail with the exact error from #4009:

```
[ERROR] Issue4009Test.testClassArrayField
        java.lang.ClassFormatError: Illegal class name "com/alibaba/fastjson2/writer/OWG_2_0_Class[]"
```

Full suites green locally on JDK 26 — core 7965, extension 56, fastjson1-compatible 1355, kotlin 36, safemode-test 2 — and core again with `-Dfastjson2.creator=reflect`.

Closes #4009

<details>
<summary>中文说明</summary>

## 问题

Bean 中含有数组类型字段时，`JSON.toJSON(bean)` 抛 `ClassFormatError`：

```java
public class BeanWithClassArray {
    private String name;
    private Class<?>[] classes;
    // getters / setters
}

JSON.toJSON(bean);
```

```
java.lang.ClassFormatError: Illegal class name "com/alibaba/fastjson2/writer/OWG_5_0_Class[]"
        in class file com/alibaba/fastjson2/writer/OWG_5_0_Class[]
```

## 根因

`ObjectWriterCreatorASM` 会把 `objectClass.getSimpleName()` 拼进它生成的类名：

```java
String className = "OWG_" + seed.incrementAndGet() + "_" + fieldWriters.size()
        + (objectClass == null ? "" : ("_" + objectClass.getSimpleName()));
```

数组类型的 simple name 以 `[]` 结尾（`Class[]`），而 `[` 在类名中非法，`defineClass` 因此拒绝生成的字节码。

## 修复

拼接前先对 simple name 做净化：数组记法转成后缀（`Class[]` → `ClassArray`），其余在类名中非法的字符替换为下划线。

方法放在 `com.alibaba.fastjson2.internal.asm.ASMUtils` —— `ObjectWriterCreatorASM` 和 `ObjectReaderCreatorASM` 本来就 static import 了它，两边不会各自演化：

```java
public static String sanitizeClassName(String simpleName)
```

### 影响范围

只有 writer 会产生非法类名。reader 那两处是防御性对称，这一点是实测确认的，不是推断：

```
reader(Class[]) -> com.alibaba.fastjson2.reader.ObjectReaderAdapter
writer(Class[]) -> com.alibaba.fastjson2.writer.OWG_1_0_ClassArray
```

`ObjectReaderCreatorASM` 对数组类型根本不生成类，直接返回 `ObjectReaderAdapter`，走不到 `ORG_` 路径。另一处（`createValueConsumer0` 里的 `VBACG_`/`VCACG_`）退出得更早：它要求有默认构造器且类是 public，数组类型两条都不满足。

## 测试

- `Issue4009Test` —— `Class[]` 与 `int[][]` 字段的序列化及 round-trip。
- `ASMUtilsTest` —— `sanitizeClassName` 的单测：null/空、无需改写的名字、数组（含 `int[][]`）、特殊字符，以及组合场景 `My$Class[]`。

## 验证

Revert-probe：把 `sanitizeClassName` 从 `ObjectWriterCreatorASM` 移除后，集成测试以 #4009 中一模一样的错误失败：

```
[ERROR] Issue4009Test.testClassArrayField
        java.lang.ClassFormatError: Illegal class name "com/alibaba/fastjson2/writer/OWG_2_0_Class[]"
```

本地 JDK 26 全量通过 —— core 7965、extension 56、fastjson1-compatible 1355、kotlin 36、safemode-test 2；`-Dfastjson2.creator=reflect` 下 core 同样通过。

</details>

### CODE REVIEW

**Comment by wenshao** [comment]
No issues found. LGTM! ✅ _— gpt-5.5 via Qwen Code /review_

File: core/src/main/java/com/alibaba/fastjson2/writer/ObjectWriterCreatorASM.java [hunk 1]
```diff
@@ -38,6 +38,39 @@ public class ObjectWriterCreatorASM
...
 
+    /**
+     * Sanitizes a class simple name for use in generated ASM class names.
+     * Array types like "Class[]" become "ClassArray", and any other invalid
+     * characters are replaced with underscores.
+     */
+    static String sanitizeClassName(String simpleName) {
+        if (simpleName == null || simpleName.isEmpty()) {
+            return "";
+        }
```
Comment (wenshao) [inline core/src/main/java/com/alibaba/fastjson2/writer/ObjectWriterCreatorASM.java]: **[Suggestion]** The 33-line `sanitizeClassName` method is duplicated verbatim in `ObjectWriterCreatorASM` and `ObjectReaderCreatorASM`, along with identical unit tests in two separate test files. Both classes already `import static com.alibaba.fastjson2.internal.asm.ASMUtils.*`, which is a natural shared home. — Concrete cost: a future edge-case fix applied to one copy but not the other causes serialization to work while deserialization still throws `ClassFormatError` (or vice versa), making the bug appear environment-specific.

```suggestion
    // Move to com.alibaba.fastjson2.internal.asm.ASMUtils:
    public static String sanitizeClassName(String simpleName) {
        // ... (single shared implementation)
    }
    // Both creators delegate: ASMUtils.sanitizeClassName(objectClass.getSimpleName())
```

_— qwen3.8-max-preview via Qwen Code /review_

**Review by wenshao — COMMENTED** [review]
Reviewed. Suggestions are inline.

_— qwen3.8-max-preview via Qwen Code /review_

File: core/src/test/java/com/alibaba/fastjson2/issues_4000/Issue4009Test.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    @Test
+    public void testClassArrayField() {
+        // Test that a class with a Class[] field can be serialized without ClassFormatError
+        BeanWithClassArray bean = new BeanWithClassArray();
+        bean.setClasses(new Class[]{String.class, Integer.class});
+        bean.setName("test");
+
+        // This should not throw ClassFormatError
+        Object json = JSON.toJSON(bean);
+        assertNotNull(json);
```
Comment (wenshao) [inline core/src/test/java/com/alibaba/fastjson2/issues_4000/Issue4009Test.java:22]: **[Suggestion]** Integration test only exercises the writer (serialization) path. The two changed `sanitizeClassName` call sites in `ObjectReaderCreatorASM` (deserialization) have no integration test. — Concrete cost: if the reader's ASM class-generation path has a constraint the writer does not, a `ClassFormatError` during deserialization would ship undetected.

```suggestion
        // This should not throw ClassFormatError
        Object json = JSON.toJSON(bean);
        assertNotNull(json);

        // Round-trip: verify deserialization also works
        String jsonString = JSON.toJSONString(bean);
        BeanWithClassArray result = JSON.parseObject(jsonString, BeanWithClassArray.class);
        assertNotNull(result);
        assertEquals("test", result.getName());
```

_— qwen3.8-max-preview via Qwen Code /review_

File: core/src/main/java/com/alibaba/fastjson2/writer/ObjectWriterCreatorASM.java (hunk 1 above)
Comment (wenshao) [reply core/src/main/java/com/alibaba/fastjson2/writer/ObjectWriterCreatorASM.java]: Done — `sanitizeClassName` now lives in `com.alibaba.fastjson2.internal.asm.ASMUtils` as a single public method, and both creators reach it through the `import static ASMUtils.*` they already had, so the call sites are unchanged apart from dropping the class qualifier.

The two verbatim-duplicated test classes (`ObjectWriterCreatorASMTest` / `ObjectReaderCreatorASMTest`, identical except for the class name) are deleted; their cases moved into the existing `ASMUtilsTest`, with `My$Class[]` added to cover the array and special-character branches combined.

File: core/src/test/java/com/alibaba/fastjson2/issues_4000/Issue4009Test.java (hunk 2 above)
Comment (wenshao) [reply core/src/test/java/com/alibaba/fastjson2/issues_4000/Issue4009Test.java:22]: Added the round-trip, but it does not gate the reader change, and I think the honest answer is that nothing can — the premise turned out not to hold.

Revert-probe, reader call sites restored to `objectClass.getSimpleName()`: `Issue4009Test` still passes, 3/3. So I checked what the two creators actually do with an array class:

```
reader(Class[]) -> com.alibaba.fastjson2.reader.ObjectReaderAdapter
writer(Class[]) -> com.alibaba.fastjson2.writer.OWG_1_0_ClassArray
```

The reader never generates a class for an array type — it returns `ObjectReaderAdapter`, so the `ORG_` path is not reached. The other reader call site (`VBACG_`/`VCACG_` in `createValueConsumer0`) bails out earlier still: it requires a default constructor and a public class, and an array type has neither. The writer is the only path that produced the illegal name, which matches #4009 (`OWG_5_0_Class[]`).

So the reader-side call is defensive symmetry, not a fix for a reachable bug, and I kept it for exactly the drift reason you gave in the other comment. The behaviour that is genuinely gated:

- writer, revert-probed — `sanitizeClassName` removed from `ObjectWriterCreatorASM` makes `Issue4009Test.testClassArrayField` fail with `ClassFormatError: Illegal class name "com/alibaba/fastjson2/writer/OWG_2_0_Class[]"`;
- the sanitizing logic itself — the unit tests now in `ASMUtilsTest`.

The round-trip assertions are still worth keeping as regression coverage (`Class[]` needs `JSONReader.Feature.SupportClassForName`, since the field type is `Class[]`; `int[][]` round-trips as-is), so I left them in — just not claiming they cover the reader's class-name generation.

Also rebased onto current `main` (the branch was 33 commits behind). Green locally on JDK 26: core 7965 + extension 56 + fastjson1-compatible 1355 + kotlin 36 + safemode-test 2, and core again with `-Dfastjson2.creator=reflect`.
